pkg/action/executor/horizontal_scaler.go

```diff
@@ -0,0 +1,308 @@
+			return nil, err
+		}
+		providerPod = foundPod
+	} else if targetEntityType == proto.EntityDTO_VIRTUAL_APPLICATION {
+		// TODO, unbind action is send as MOVE
+		// An example actionItemDTO is as follows:
+
+		// actionItemDTO:<actionType:MOVE uuid:"_YaxeQDPxEea2efRvkwVM-Q"
+		// targetSE:<entityType:VIRTUAL_APPLICATION id:"vApp-apache2" displayName:"vApp-apache2" > currentSE:<
+		// entityType:APPLICATION id:"apache2::default/frontend-hhp6g" displayName:"apache2::default/frontend-
+		// hhp6g" > providers:<entityType:APPLICATION ids:"apache2::default/frontend-57ytg" ids:"apache2::
+		// default/frontend-ue0kg" ids:"apache2::default/frontend-hhp6g" ids:"apache2::default/frontend-luohe">>
+		//
+		// TODO find the pod name based on application ID. App id is in the following format.
+		// TODO, maybe use service related info to find the corresponding pod.
+		// ProcessName::PodNamespace/PodName
+		currentSE := actionItem.GetCurrentSE()
+		targetEntityType := currentSE.GetEntityType()
+		if targetEntityType != proto.EntityDTO_APPLICATION {
+
+		}
+		appName := currentSE.GetId()
+		ids := strings.Split(appName, "::")
+		if len(ids) < 2 {
+			return nil, fmt.Errorf("%s is not a valid Application ID. Unbind failed.", appName)
+		}
+		podIdentifier := ids[1]
+
+		// Get the target pod from podIdentifier.
+		pod, err := util.GetPodFromIdentifier(h.kubeClient, podIdentifier)
+		if err != nil {
+			return nil, err
+		}
+		providerPod = pod
+	}
+	return providerPod, nil
+}
+
+func (h *HorizontalScaler) horizontalScale(action *turboaction.TurboAction) (*turboaction.TurboAction, error) {
+	// 1. Setup consumer
+	actionContent := action.Content
+	scaleSpec, ok := actionContent.ActionSpec.(turboaction.ScaleSpec)
+	if !ok || scaleSpec.NewReplicas < 0 {
+		return nil, errors.New("Failed to setup horizontal scaler as the provided scale spec is invalid.")
+	}
+
+	var key string
+	if actionContent.ParentObjectRef.ParentObjectUID != "" {
+		key = actionContent.ParentObjectRef.ParentObjectUID
+	} else {
+		return nil, errors.New("Failed to setup horizontal scaler consumer: failed to retrieve the key.")
+	}
+	glog.V(3).Infof("The current horizontal scaler consumer is listening on the key %s", key)
+	podConsumer := turbostore.NewPodConsumer(string(action.UID), key, h.broker)
+
+	// 2. scale up and down by changing the replica of replication controller or deployment.
+	err := h.updateReplica(actionContent.TargetObject, actionContent.ParentObjectRef, actionContent.ActionSpec)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to update replica: %s", err)
+	}
+
+	// 3. If this is an unbind action, it means it is an action with only one stage.
+	// So after changing the replica it can return immediately.
+	if action.Content.ActionType == turboaction.ActionUnbind {
+		// Update turbo action.
+		action.Status = turboaction.Executed
+		return action, nil
+	}
+
+	// 4. Wait for desired pending pod
+	// TODO, should we always block here?
+	pod, ok := <-podConsumer.WaitPod()
```

Comment:
We need to think of a timeout mechanism to make sure it's not waiting forever. For example, what if there's other scheduler picks up the pending pod?